sway/desktop/xdg_shell.c

```diff
@@ -66,14 +66,28 @@ static void popup_handle_destroy(struct wl_listener *listener, void *data) {
 static void popup_unconstrain(struct sway_xdg_popup *popup) {
 	struct sway_view *view = popup->child.view;
 	struct wlr_xdg_popup *wlr_popup = popup->wlr_xdg_surface->popup;
+	struct wlr_box toplevel_geometry = {};
+	struct wlr_surface *parent = wlr_popup->parent;
 
 	struct sway_output *output = view->container->pending.workspace->output;
 
+	while (wlr_surface_is_xdg_surface(parent)) {
+		struct wlr_xdg_surface *xdg_surface =
+			wlr_xdg_surface_from_wlr_surface(parent);
+
+		if (xdg_surface->role == WLR_XDG_SURFACE_ROLE_POPUP) {
+			parent = xdg_surface->popup->parent;
+		} else {
+			toplevel_geometry = xdg_surface->pending.geometry;
+			break;
+		}
+	}
+
 	// the output box expressed in the coordinate system of the toplevel parent
 	// of the popup
 	struct wlr_box output_toplevel_sx_box = {
-		.x = output->lx - view->container->pending.content_x,
-		.y = output->ly - view->container->pending.content_y,
+		.x = output->lx - view->container->pending.content_x + toplevel_geometry.x,
+		.y = output->ly - view->container->pending.content_y + toplevel_geometry.y,
```

Comment:
Can we use `pending.x` instead? Or is there a difference?

Reply:
You mean `view->container->pending.x`? That is a container position, it's not related to the surface or surface geometry. Just view + borders.

There's `view->container->view->geometry` which contains the right values, but I don't know if it's guaranteed to be a toplevel view or if that works with transactions.